In [9]:
# #2 Yelp API
# Get set up on the Yelp Fusion API. I've put some details on how to do that on the page on a Yelp Authentication setup page which will likely be very useful!
# Install the yelpapi module which is online: there's both a documentation page on the the Python Package Index (PyPI) website and a Github page with some documentation. As I said in class, you can either do this by (a) opening a terminal on your system and running pip install yelpapi or you can try running pip install yelpapi in your Jupyter notebook. Reach on out teams or in open lab sessions if you run into trouble.
# Create a new .py file (e.g., I called mine yelp_authentication.py) in the same directory as your Yelp notebooks are using and add your API key to it. Then use the import command to use your API key in a notebook without having the key itself visible in the notebook!
# Once you've done this, use your yelp data collection notebook to grab a list of 50 businesses of any kind (your choice!), in any city (again, your choice!) using Yelp and the yelpapi module. This should be easy if you modify the notebook from the Community Data Science Course (Spring 2023)/Week 6 lecture.
# Once you have done this, add some code so that you save the "raw" JSON output to a .json or .jsonl file (whichever is appropriate).
# Now create a second notebook that opens up that file, reads the data, and outputs a TSV file with the the name of the business, the average rating, and at least three other pieces or metadata that are available in the Yelp API

import yelp_authentication;
import requests
import json
import yelpapi

def do_yelp_request(term):
    headers = {
    'Authorization': 'Bearer ' + yelp_authentication.yelp_api_key,
    }

    url = 'https://api.yelp.com/v3/businesses/search'
# 3. Define my parameters of the search
#   In this example, we're asking for:
#   up to 50 places ('limit') that match pasta ('term')
#   within 10000m ('radius') of Seattle ('location').
    params = {
        'location': 'Chicago',
        'term': term,
        'radius': 10000,
        'limit': 50,
    }

    response = requests.get(url, headers=headers, params=params)
    print("ratelimit-remaining:"+response.headers['ratelimit-remaining'])
    if response.status_code == 200:
        data = json.loads(response.text)
        #businesses = data['businesses']
        return data        
    else:
        print('Error:', response.status_code, response.text)
        return None


In [11]:
coffee_in_chicago = do_yelp_request("coffee")

ratelimit-remaining:4996.0


In [25]:
for business in coffee_in_chicago["businesses"]:
    print(f"{business['name']}\'s rating is {business['rating']}")

A Cup of Joe 's rating is 4.5
1321 Coffee Shop's rating is 5.0
Sputnik Coffee Company's rating is 5.0
Cafe El Meson's rating is 5.0
Mandala Cafe's rating is 5.0
Bad Owl Coffee Roasters's rating is 4.5
The Wormhole Coffee's rating is 4.5
La Catedral Cafe & Restaurant's rating is 4.5
Starbucks's rating is 2.5
Muse Coffee Studio's rating is 5.0
Tasa Coffee Roasters's rating is 5.0
Brewpoint Craft Oak Park's rating is 4.5
The Club house's rating is 5.0
Avito Caffe's rating is 4.5
Dulce Mami Cafe's rating is 3.5
Beelove Cafe's rating is 5.0
Sawada Coffee's rating is 4.5
Big Shoulders Coffee's rating is 2.0
La Borra Del Cafe's rating is 4.0
Foxglove Coffee's rating is 4.5
El Churro Shop's rating is 4.0
Kafecito's rating is 3.0
Friendly Coffee Lounge's rating is 5.0
Philosophy Coffee Roasting Facility & Offices's rating is 5.0
Masa Madre's rating is 5.0
L Cafecito Jumping Bean's rating is 4.5
Green Tomato Cafe's rating is 4.5
Whirlwind Coffee Company's rating is 5.0
Netccino Internet Cafe's r

In [27]:
# Once you have done this, add some code so that you save the "raw" JSON output to a .json or .jsonl file (whichever is appropriate).
def create_json_file(file_name_json, jsondata):
    with open(file_name_json+".json", "w") as my_file:
        json_string = json.dumps(jsondata) #converts into json string        
        print(json_string, file=my_file) #This was saved as a long string, not as spreadsheet as we had on the past.
    
    print(f"{file_name_json} created")

file_name = "yelp_coffee_in_chicago"
create_json_file(file_name,coffee_in_chicago)

yelp_coffee_in_chicago created


In [29]:
# Now create a second notebook that opens up that file, reads the data, and outputs a TSV file with the the name of the business, the average rating, and at least three other pieces or metadata that are available in the Yelp API

def open_json_to_dict(file_name):
    with open(file_name+".json", 'r') as input_file:
        input_data = input_file.read()
        new_data = json.loads(input_data) #load into new_data so we won't have to work as big string.
        return new_data

new_data= open_json_to_dict(file_name)

In [31]:
print(new_data["businesses"][0])

with open(file_name+".tsv", "w") as f:
        print("name\trating\tprice\tphone\tcategories", file=f)
        
        for business in new_data["businesses"]:
            #print(business)
            name = business["name"]
            average_rating = business["rating"]
            price = ''
            if "price" in business:
                price = business['price']
            phone = business["display_phone"]
            categories = business["categories"]

            print(name, "\t", average_rating, "\t", price, "\t", phone, "\t", categories, file=f)
            
print(f"{file_name}.tsv created")

{'id': 'zSoZYT0PcBZYkG-jkKT-vw', 'alias': 'a-cup-of-joe-chicago', 'name': 'A Cup of Joe ', 'image_url': 'https://s3-media1.fl.yelpcdn.com/bphoto/HDnjXWkg9ErjaqViE1tlcA/o.jpg', 'is_closed': False, 'url': 'https://www.yelp.com/biz/a-cup-of-joe-chicago?adjust_creative=XVda9xQ_42xVfoeMInQ5Fw&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=XVda9xQ_42xVfoeMInQ5Fw', 'review_count': 96, 'categories': [{'alias': 'coffee', 'title': 'Coffee & Tea'}], 'rating': 4.5, 'coordinates': {'latitude': 41.7923651166731, 'longitude': -87.7919535711408}, 'transactions': ['pickup', 'delivery'], 'price': '$', 'location': {'address1': '6806 W Archer Ave', 'address2': '', 'address3': None, 'city': 'Chicago', 'zip_code': '60638', 'country': 'US', 'state': 'IL', 'display_address': ['6806 W Archer Ave', 'Chicago, IL 60638']}, 'phone': '+17733060185', 'display_phone': '(773) 306-0185', 'distance': 6683.692992213431}
yelp_coffee_in_chicago.tsv created


# What is your proposed unit of analysis? In other words, if/when you end up building something like a spreadsheet, what are rows going to represent?
I want to create graphics and a spreadsheet to display the engagement of each TV series. I have decided to focus on TV series, as HBO does not produce original movies; rather, they produce documentaries that fall into specific genres of movies. If I were to include only movies, the data would not be as accurate. The graphics will illustrate which platform has more engagement by measuring the number of viewers for each original series.



# What specific measures associated with each unit do you want to collect? In other words, what are the columns in the spreadsheet going to be?
I will be tracking the number of viewers per episode for each series. As I have done more research, I have become aware that obtaining production cost information will be difficult as it is highly restrictive. Therefore, I will only be using engagement information to predict which platform has more engagement, resulting in more long-term subscribers.


# Tell us what you've learned about the API:
Are you able to get the data you want with one API call or many? If more than one, how many?
I will be using 2 APIs the first one Wikipedia and the second one Fandom.com
If it's more than one call, how will you know when you have collected all your data?
Just one call for entry is going to be required for the APIs.


# API call

In [33]:
import requests
import json
# import time

wp_api_url = "http://en.wikipedia.org/w/api.php/"
parameters = {'action' : 'query',
                  'titles' : "The_Last_of_Us_(TV_series)",
                  'prop' : 'revisions',
                  'rvprop' : 'flags|timestamp|user|size|ids',
                  'rvlimit' : 500,
                  'format' : 'json',
                   }
api_answers =[]
while True:
        # this will wait for one second
        # time.sleep(1)
        
        # the first line open the urls but also handles unicode urls
        call = requests.get(wp_api_url, params=parameters)
        api_answer = call.json()
        
        # now we'll add this to whatever we are tracking
        api_answers.append(api_answer)
        
        # 'continue' tells us there's more revisions to add
        if 'continue' in api_answer.keys():
            # replace the 'continue' parameter with the contents of the
            # api_answer dictionary.
            parameters.update(api_answer['continue'])
        else:
            break

print(api_answers)

[{'continue': {'rvcontinue': '20230208142258|1138196680', 'continue': '||'}, 'query': {'normalized': [{'from': 'The_Last_of_Us_(TV_series)', 'to': 'The Last of Us (TV series)'}], 'pages': {'63292072': {'pageid': 63292072, 'ns': 0, 'title': 'The Last of Us (TV series)', 'revisions': [{'revid': 1153742534, 'parentid': 1153738842, 'user': 'SibTower1987', 'timestamp': '2023-05-08T02:15:43Z', 'size': 281001}, {'revid': 1153738842, 'parentid': 1152962751, 'user': 'Rhain', 'timestamp': '2023-05-08T01:45:13Z', 'size': 280967}, {'revid': 1152962751, 'parentid': 1152829011, 'user': 'Citation bot', 'timestamp': '2023-05-03T11:49:08Z', 'size': 282092}, {'revid': 1152829011, 'parentid': 1152828225, 'user': 'Rhain', 'timestamp': '2023-05-02T15:26:51Z', 'size': 282083}, {'revid': 1152828225, 'parentid': 1152827915, 'user': 'Indagate', 'timestamp': '2023-05-02T15:20:56Z', 'size': 281774}, {'revid': 1152827915, 'parentid': 1152827651, 'minor': '', 'user': 'Rhain', 'timestamp': '2023-05-02T15:18:27Z', '

# How big is the JSON file that you saved on your disk (i.e., in bytes or kilobytes)? If it is not your full dataset, what is your estimate for how much larger the full dataset will be? How big will the total dataset be? Is that a problem?
I dont´t have my full final dataset set yet but the test request is about 236 KB. I plan to analyse 10 series so it should be about 2.5 MB for one API.
